Add apply command

lucperkins opened the pull request:
The `apply` command resolves a reference into a store path and applies it to the specified Nix profile:

```shell
fh apply nixos "omnicorp/web/*#nixosConfigurations.auth-server"
```

corresponds to

```shell
nix build \
  --print-build-logs \
  --max-jobs 0 \
  --profile /nix/var/nix/profiles/system \
  $(fh resolve "omnicorp/web/*#nixosConfigurations.auth-server")

/nix/var/nix/profiles/system/bin/switch-to-configuration switch
```

cole-h commented on src/cli/cmd/convert.rs:
```diff
@@ -81,12 +81,8 @@ impl CommandExecute for ConvertSubcommand {
-                .status()
-                .await?;
+
+            nix_command(&["flake", "lock"]).await?;
```
(don't forget to wrap with context so that if this fails we know why)

cole-h commented on src/cli/cmd/resolve.rs:
```diff
@@ -38,12 +42,24 @@ impl CommandExecute for ResolveSubcommand {
+                "--print-build-logs",
+                &resolved_path.store_path,
+            ])
+            .await?;
```
here too

cole-h commented on src/cli/cmd/resolve.rs:
```diff
@@ -16,6 +16,10 @@ pub(crate) struct ResolveSubcommand {
     #[arg(long)]
     json: bool,
 
+    /// Build the resolved path with Nix.
+    #[arg(short, long, default_value_t = false)]
+    build: bool,
```
This isn't even correct. Since you pass `--max-jobs 0`, no building will happen. It'll either substitute the path (if it exists), or fail. Maybe `--fetch` or `--realise` (British spelling on purpose to match the rest of the Nix CLI) would be a better flag.

lucperkins replied:
Switched to `--fetch` but I'm open to suggestion

cole-h commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +368,20 @@ macro_rules! flakehub_url {
 }
 
+async fn nix_command(args: &[&str]) -> Result<(), FhError> {
+    if !command_exists("nix") {
+        return Err(FhError::MissingExecutable(String::from("nix")));
+    }
```
Why not make `command_exists` return this error itself? Then it becomes something like `check_command_exists("bin")?;`

grahamc commented on flake.nix:
```diff
@@ -103,7 +103,7 @@
             packages = with pkgs; [
               (fenixToolchain system)
-              cargo-watch
+              bacon
```
I rather like cargo watch?

lucperkins replied:
I promise you'll like bacon more. But I can keep them both.

grahamc reviewed (dismissed):
It is actually not safe to nix-build without a symlink or profile updated. I think we probably want a verby way to say like:

```
fh profile system pull Foo/bar/v0#mything
```

...but I'm not sure what the command should be. Let's brainstorm?

lucperkins commented:
@grahamc I've changed it up so it works like this:

```shell
fh resolve --profile system "omnicorp/systems/*#nixosConfigurations.auth-server"
```

lucperkins commented:
@grahamc Okay, I tried something different here:

```shell
fh apply system "omnicorp/systems/*#nixosConfigurations.auth-server"
```

grahamc reviewed (commented):
This is looking good to me. I'll give it a test.

lucperkins commented:
@grahamc I've updated this to detect `switch-to-configuration` and also added an optional `--switch` flag that actually runs it.

```shell
# Runs it for you
fh apply system --switch "omnicorp/systems/*#nixosConfigurations.auth-server"

# Displays helpful info about how to run it
fh apply system "omnicorp/systems/*#nixosConfigurations.auth-server"
```

lucperkins commented:
@grahamc I've changed this up a bit as an experiment. This now works:

```shell
fh apply nixos --run switch "omnicorp/systems/*#nixosConfigurations.auth-server"
```

Curious what you think.

grahamc commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,113 @@ macro_rules! flakehub_url {
+        );
+    }
+
+    let output = if use_sudo {
+        tokio::process::Command::new("sudo")
+            .args(["nix", "--extra-experimental-features", "nix-command flakes"])
+            .args(args)
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn()
+            .wrap_err("failed to spawn Nix command")?
+            .wait_with_output()
+            .await
+            .wrap_err("failed to wait for Nix command output")?
+    } else {
+        tokio::process::Command::new("nix")
+            .args(["--extra-experimental-features", "nix-command flakes"])
+            .args(args)
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn()
+            .wrap_err("failed to spawn Nix command")?
+            .wait_with_output()
+            .await
+            .wrap_err("failed to wait for Nix command output")?
+    };
```
If you'd like, this can be deduped like this:

```rust
    let mut cmd = if use_sudo {
        let mut cmd = tokio::process::Command::new("sudo");
        cmd.arg("nix");
        cmd
    } else {
        tokio::process::Command::new("nix")
    };
    
    cmd.args(["--extra-experimental-features", "nix-command flakes"]);
    cmd.args(args);
    cmd.stdout(Stdio::inherit());
    cmd.stderr(Stdio::inherit());

    let output = cmd
            .spawn()
            .wrap_err("failed to spawn Nix command")?
            .wait_with_output()
            .await
            .wrap_err("failed to wait for Nix command output")?;
```

lucperkins replied:
Good call

grahamc commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,113 @@ macro_rules! flakehub_url {
+
+fn parse_flake_output_ref(output_ref: &str) -> Result<FlakeOutputRef, FhError> {
+    // Ensures that users can use both forms:
+    // 1. https://flakehub/f/{org}/{project}/{version_req}#{output}
```
nice

grahamc commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,113 @@ macro_rules! flakehub_url {
+
+// Ensure that release refs are of the form {org}/{project}/{version_req}
+fn parse_release_ref(flake_ref: &str) -> Result<String, FhError> {
+    match flake_ref.split('/').collect::<Vec<_>>()[..] {
+        [org, project, version_req] => {
+            validate_segment(org)?;
+            validate_segment(project)?;
+            validate_segment(version_req)?;
+
+            Ok(String::from(flake_ref))
+        }
+        _ => Err(FhError::FlakeParse(format!(
+            "flake ref {flake_ref} invalid; must be of the form {{org}}/{{project}}/{{version_req}}"
+        ))),
```
hell yeah

lucperkins replied:
Pattern matching 🚀

grahamc commented on src/cli/cmd/resolve.rs:
```diff
@@ -13,7 +13,7 @@ pub(crate) struct ResolveSubcommand {
 
     /// Output the result as JSON displaying the store path plus the original attribute path.
-    #[arg(long)]
+    #[arg(long, env = "FH_JSON_OUTPUT")]
```
```suggestion
    #[arg(long, env = "FH_OUTPUT_JSON")]
```

in case we have more outputs later?

grahamc replied:
(some in the next file)

grahamc commented on src/cli/cmd/apply/mod.rs:
```diff
@@ -0,0 +1,200 @@
+            &profile_path,
+            store_path,
+        ],
+        true,
```
iirc this `true` means use sudo, but we shouldn't use sudo for `home-manager` -- let's hoist this up to a parameter based on the system type

grahamc commented on src/cli/cmd/apply/mod.rs:
```diff
@@ -0,0 +1,200 @@
+        match self.system {
+            System::HomeManager(_) => {
+                // /nix/store/{path}/activate
+                let script_path = path!(&resolved_path.store_path, HOME_MANAGER_SCRIPT);
```
I think elsewhere we're downloading the closure in `apply_path_to_profile`, which we're not running here. We don't run it here, because HomeManager takes care of it. But I think that means we haven't actually downloaded it in the HM case.

grahamc commented on src/cli/cmd/apply/mod.rs:
```diff
@@ -0,0 +1,200 @@
+    /// Resolve the store path for a Home Manager configuration and run its activation script
+    HomeManager(HomeManager),
+
+    #[cfg(target_os = "macos")]
+    /// Resolve the store path for a nix-darwin configuration and run its activation script
+    NixDarwin(NixDarwin),
+
+    #[cfg(target_os = "linux")]
+    /// Apply the resolved store path on a NixOS system
+    #[clap(name = "nixos")]
+    NixOs(NixOs),
```
If we can avoid `cfg(target_os...`, I'd like to. We recently deleted a ton of this from the Nix installer project because it makes developing on different platforms much much harder.

grahamc commented on src/cli/cmd/apply/mod.rs:
```diff
@@ -0,0 +1,200 @@
+            System::NixDarwin(nix_darwin) => nix_darwin.output_ref()?,
+        };
+
+        tracing::info!("Resolving store path for output: {}", output_ref);
```
```suggestion
        tracing::info!("Resolving {}", output_ref);
```

grahamc commented on src/cli/cmd/apply/nix_darwin.rs:
```diff
@@ -0,0 +1,76 @@
+//
+// fh apply nix-darwin omnicorp/home/0.1
+//
+// Here, `omnicorp/systems/0.1` resolves to `omnicorp/systems/0#darwinConfigurations.$(devicename).system`.
```
```suggestion
// Here, `omnicorp/systems/0.1` resolves to `omnicorp/systems/0#darwinConfigurations.$(devicename)`.
```

grahamc commented on src/cli/cmd/apply/nix_darwin.rs:
```diff
@@ -0,0 +1,76 @@
+    /// is the output of scutil --get LocalHostName.
+    pub(super) output_ref: String,
+
+    #[arg(long, short, env = "FH_APPLY_PROFILE", default_value = "system")]
+    pub(super) profile: String,
```
Vetted: this does in fact default to the `system` profile:

```
profile=/nix/var/nix/profiles/system                                                                                                                         action=                                                                                                                                                      flake=
```

grahamc commented on src/cli/cmd/apply/home_manager.rs:
```diff
@@ -0,0 +1,74 @@
+    /// If the latter, the attribute path defaults to homeConfigurations.{whoami}.
+    pub(super) output_ref: String,
+
+    #[arg(long, short, env = "FH_APPLY_PROFILE", default_value = "system")]
+    pub(super) profile: String,
```
Home manager doesn't use the `system` profile, or even a profile in `/nix/var/nix/profiles/*` at all. In any case we probably shouldn't set this profile ourselves -- since home manager takes pains to do something specific here.

I think for home-manager we shouldn't be installing to a profile at all, but perhaps using a NamedTemporaryDirectory that we write to -- do the action they request -- and then delete.

cole-h commented on src/cli/cmd/apply/mod.rs:
```diff
@@ -0,0 +1,225 @@
+                )
+                .await?;
+
+                // /nix/store/{path}/activate
+                let script_path = path!(&profile_path, HOME_MANAGER_SCRIPT);
+
+                run_script(script_path, None, HOME_MANAGER_SCRIPT).await?;
+            }
+            System::NixDarwin(NixDarwin { profile, .. }) => {
+                apply_path_to_profile(
+                    Some(&profile),
+                    &resolved_path.store_path,
+                    true, // sudo if necessary when running `nix build`
+                )
+                .await?;
+
+                // {path}/sw/bin/darwin-rebuild
+                let script_path = path!(&profile, "sw", "bin", NIX_DARWIN_SCRIPT);
+
+                run_script(
+                    script_path,
+                    Some(DARWIN_REBUILD_ACTION.to_string()),
+                    NIX_DARWIN_SCRIPT,
+                )
+                .await?;
+            }
+            System::NixOs(NixOs { action, .. }) => {
+                let profile_path = Path::new(NIXOS_PROFILE);
+                apply_path_to_profile(
+                    Some(Path::new(NIXOS_PROFILE)),
+                    &resolved_path.store_path,
+                    true, // sudo if necessary when running `nix build`
+                )
+                .await?;
+
+                let script_path = path!(&profile_path, "bin", NIXOS_SCRIPT);
+
+                run_script(script_path, Some(action.to_string()), NIXOS_SCRIPT).await?;
+            }
+        }
```
Is there a reason this couldn't be something implemented on the inner types and passed through the enum? So it would become something like `self.system.apply_path(&resolved_path.store_path)`

grahamc replied:
Do you mean like make a trait and use it here?

lucperkins replied:
Even with a commonly implemented trait, I'm pretty sure you'd still have to call the method on each variant, like we do with `CommandExecute`

cole-h replied:
It doesn't have to be a trait (though that would make it slightly easier) -- just some function implemented on the `System` enum that calls into specialized functions in the inner structs.

cole-h commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,109 @@ macro_rules! flakehub_url {
+    // 2. {org}/{project}/{version_req}#{output}
+    let output_ref = String::from(
+        output_ref
+            .strip_prefix("https://flakehub.com/f/")
```
How does this work with dev instances?

grahamc replied:
It's just stripping the prefix, so I suppose in dev I think you'd probably use org/project/...  or if you used prod URLs it'd get stripped down to org/project/.... But if you used staging prefixes it won't ... which ... what would you like? :)

cole-h replied:
I would like it to work the same for both dev and prod URLs so that it's possible to test it outside of prod.

lucperkins replied:
I just boarded a plane but `api_addr` is globally available and could be used to construct the prefix

grahamc replied:
Fixed up!

cole-h commented on .github/workflows/rust.yml:
```diff
@@ -6,16 +6,38 @@ on:
         run: nix develop --command cargo fmt --check
 
+  rust-clippy-and-test:
+    # We need to run the checks on Linux and macOS because there are system-specific
+    # versions of fh (e.g. `fh apply nix-darwin` is only on macOS while `fh apply
+    # nixos` is only on Linux)
+    strategy:
```
I don't think this is true anymore, so let's combine them back up and rename it back to `rust-fmt-and-clippy`

grahamc replied:
Like just revert the chanegs to this file?

cole-h replied:
Pretty much, except I imagine we'd want to keep the updated `actions/checkout`

cole-h commented on src/cli/cmd/apply/mod.rs:
```diff
@@ -0,0 +1,275 @@
+
+        if let Ok(script_path_metadata) = tokio::fs::metadata(&script_path).await {
+            let permissions = script_path_metadata.permissions();
+            if permissions.mode() & 0o111 != 0 {
```
Can I get a brief comment on what this is checking? And if the check fails, can we have some sort of logging to let the user know that something was in an unexpected state?

cole-h commented on src/cli/cmd/apply/nix_darwin.rs:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    fn default_ref(&self) -> String {
+        format!("darwinConfigurations.{}", whoami::devicename(),)
```
nit

```suggestion
        format!("darwinConfigurations.{}", whoami::devicename())
```

cole-h commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,99 @@ macro_rules! flakehub_url {
+            args.join(" ")
+        );
+
+        let mut cmd = tokio::process::Command::new("sudo");
```
I wonder if we'll run into issues where the user's sudo config does / doesn't forward some environment variables.

Something to be aware of, I guess. (Maybe we should allow the `sudo` command to be customized, so that users have an escape hatch.... but I guess we can revisit that if/when people have issues.)

grahamc replied:
Let's revisit on error.

cole-h commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,99 @@ macro_rules! flakehub_url {
+            "Current user is {} rather than root; running Nix command using sudo",
+            whoami::username()
+        );
+        tracing::debug!(
+            "Running: sudo nix --extra-experimental-features 'nix-command-flakes' {}",
+            args.join(" ")
+        );
```
IMO, we should use the debug/display (I forget which it has) implementation of `std::process::Command` right before we spawn it. That way it will always contain every argument we pass.

(Tokio has a helper `.as_std()` that you can use to get the display/debug impl of the std type)

cole-h commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,99 @@ macro_rules! flakehub_url {
+    if output.status.success() {
+        Ok(())
+    } else {
+        Err(FhError::FailedNixCommand)
```
I think we should include the nix command we ran in the error message; otherwise, it's only visible in debug levels.

grahamc commented on src/cli/cmd/mod.rs:
```diff
@@ -367,6 +382,93 @@ macro_rules! flakehub_url {
+    cmd.args(args);
+    cmd.stdout(Stdio::inherit());
+    cmd.stderr(Stdio::inherit());
+    let cmd_str = format!("{:?}", cmd);
```
Yessir!

grahamc commented:
Going to run through and manually test each variant once.
